File: packages/artlavka_core/lib/utils/money.dart
/// UZS money helpers.
///
/// Money in ART-LAVKA is ALWAYS an `int` of Uzbekistani so'm with no decimals
/// (see SPEC §5). This class only *formats* for display — it never rounds or
/// does arithmetic that could lose value. Do math on the raw `int`.
abstract final class Money {
  /// Non-breaking space (U+00A0) thousands separator, so a price never wraps
  /// mid-number. Written as an escape so the source stays pure ASCII.
  static const String groupSeparator = ' ';

  /// Default currency suffix shown to users.
  static const String suffix = "so'm";

  /// Group digits, e.g. `1234567` -> `1 234 567`.
  ///
  /// Negative amounts keep the sign in front.
  static String group(int amount) {
    final negative = amount < 0;
    final digits = amount.abs().toString();
    final buffer = StringBuffer();
    for (var i = 0; i < digits.length; i++) {
      if (i != 0 && (digits.length - i) % 3 == 0) buffer.write(groupSeparator);
      buffer.write(digits[i]);
    }
    return negative ? '-${buffer.toString()}' : buffer.toString();
  }

  /// Format for display, e.g. `1234567` -> `1 234 567 so'm`.
  ///
  /// Pass [withSuffix] = false for places that render the currency separately.
  static String format(int amount, {bool withSuffix = true}) =>
      withSuffix ? '${group(amount)}$groupSeparator$suffix' : group(amount);
}

File: backend/src/common/money.ts
/**
 * Integer-UZS money helpers — the server-side mirror of the Dart `Money` util.
 *
 * Money is always `bigint` UZS with no decimals. In JSON it is serialized as a
 * STRING (see `patchBigIntJson`) so large values never lose precision crossing
 * to JS `number` / the Flutter side, which parses it back to `int`.
 */

/** Non-breaking space (U+00A0) thousands separator (written as an escape). */
export const GROUP_SEPARATOR = ' ';
export const UZS_SUFFIX = "so'm";

/** Group digits, e.g. 1234567n -> "1 234 567". */
export function groupUzs(amount: bigint): string {
  const negative = amount < 0n;
  const digits = (negative ? -amount : amount).toString();
  let out = '';
  for (let i = 0; i < digits.length; i++) {
    if (i !== 0 && (digits.length - i) % 3 === 0) out += GROUP_SEPARATOR;
    out += digits[i];
  }
  return negative ? `-${out}` : out;
}

/** Display format, e.g. "120 000 so'm". */
export function formatUzs(amount: bigint, withSuffix = true): string {
  const grouped = groupUzs(amount);
  return withSuffix ? `${grouped}${GROUP_SEPARATOR}${UZS_SUFFIX}` : grouped;
}

/** A bigint augmented with the JSON hook we install below. */
interface JsonableBigInt {
  toJSON?: () => string;
}

/**
 * Make `bigint` JSON-serializable as a string. Call once at bootstrap so any
 * response containing BigInt money serializes correctly (BACKEND_NODE.md §3/§6).
 */
export function patchBigIntJson(): void {
  (BigInt.prototype as unknown as JsonableBigInt).toJSON = function (
    this: bigint,
  ): string {
    return this.toString();
  };
}
